UnitTests/BraintreePayPalTests/BTPayPalApprovalURLParser_Tests.swift

```diff
@@ -0,0 +1,48 @@
+import XCTest
+@testable import BraintreeCore
+@testable import BraintreePayPal
+
+final class BTPayPalApprovalURLParser_Tests: XCTestCase {
+
+    func testInit_withPayPalAppApprovalUrl_setsRedirectTypeAndBAToken() {
+        let payPalAppApprovalUrlJSON = BTJSON(value: [
+            "agreementSetup": ["paypalAppApprovalUrl": "https://www.paypal.com?ba_token=A_FAKE_BA_TOKEN"]
+        ])
+        let approvalURLParser = BTPayPalApprovalURLParser(body: payPalAppApprovalUrlJSON, linkType: "universal")
+
+        XCTAssertEqual(approvalURLParser?.redirectType, .payPalApp(url: URL(string: "https://www.paypal.com?ba_token=A_FAKE_BA_TOKEN")!))
+        XCTAssertEqual(approvalURLParser?.baToken, "A_FAKE_BA_TOKEN")
+        XCTAssertNil(approvalURLParser?.ecToken)
+    }
+
+    func testInit_withRedirectUrl_setsRedirectTypeAndECToken() {
+        let redirectUrlJSON = BTJSON(value: [
+            "paymentResource": ["redirectUrl": "https://www.paypal.com/checkout?token=A_FAKE_EC_TOKEN"]
+        ])
+        let approvalURLParser = BTPayPalApprovalURLParser(body: redirectUrlJSON, linkType: nil)
+
+        XCTAssertEqual(approvalURLParser?.redirectType, .webBrowser(url: URL(string: "https://www.paypal.com/checkout?token=A_FAKE_EC_TOKEN")!))
+        XCTAssertNil(approvalURLParser?.baToken)
+        XCTAssertEqual(approvalURLParser?.ecToken, "A_FAKE_EC_TOKEN")
+    }
+
+    func testInit_withApprovalUrl_setsRedirectTypeAndECToken() {
+        let approvalUrlJSON = BTJSON(value: [
+            "agreementSetup": ["approvalUrl": "https://www.paypal.com/agreements/approve?ba_token=A_FAKE_BA_TOKEN"]
+        ])
+        let approvalURLParser = BTPayPalApprovalURLParser(body: approvalUrlJSON, linkType: "universal")
+
+        XCTAssertEqual(approvalURLParser?.redirectType, .webBrowser(url: URL(string: "https://www.paypal.com/agreements/approve?ba_token=A_FAKE_BA_TOKEN")!))
+        XCTAssertEqual(approvalURLParser?.baToken, "A_FAKE_BA_TOKEN")
+        XCTAssertNil(approvalURLParser?.ecToken)
+    }
+
+    func testInit_withBadJSON_returnsNil() {
```

Comment:
* [Test for invalid approvalURLs](https://github.com/braintree/braintree_ios/blob/f7ed71ea853aeee0cc77cf445068ab203aa864b9/UnitTests/BraintreePayPalTests/BTPayPalClient_Tests.swift#L191-L212)